io/iommu/iommu_tests.py

```diff
@@ -0,0 +1,336 @@
+#!/usr/bin/env python
+
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2022 Advanced Micro Devices, Inc.
+# Author: Dheeraj Kumar Srivastava <dheerajkumar.srivastava@amd.com>
+
+"""
+Various combination of unbind, bind, change iommu group domain type, reset
+and rescan is used to form sub-tests that test and exercise iommu code.
+"""
+
+import os
+import time
+from avocado import Test
+from avocado.utils import process
+from avocado.utils import pci, genio
+from avocado.utils.software_manager.manager import SoftwareManager
+
+
+class IommuTest(Test):
+
+    """
+    Various combination of unbind, bind, change iommu group domain type,
+    reset and rescan is used to form sub-tests that test and exercise iommu
+    code.
+
+    :param device: Name of the pci device
+    """
+
+    def setUp(self):
+        """
+        Setup the device
+        """
+        self.pci_devices = self.params.get('pci_devices', default=None)
+        self.count = int(self.params.get('count', default=1))
+        self.domains = ["DMA", "DMA-FQ", "identity", "auto"]
+        if not self.pci_devices:
+            self.cancel("No pci device given")
+        smm = SoftwareManager()
+        if not smm.check_installed("pciutils") and not smm.install("pciutils"):
+            self.cancel("pciutils package not found and installing failed")
+        # Check the number of devices in iommu-group for pci device passed.
+        for pci_addr in self.pci_devices.split(" "):
+            driver = pci.get_driver(pci_addr)
+            if driver is None:
+                self.cancel("Device passed is not bound to any driver")
+            lst = os.listdir(f'/sys/bus/pci/drivers/{driver}/{pci_addr}/'
+                             'iommu_group/devices/')
+            if len(lst) != 1:
```

Comment:
what if lenght is '0'  here?

Reply:
This will never happen as we are looking in the iommu group of pci device passed, so pci device itself be there atleast in any case.